File: app/src/main/java/org/dhis2/usescases/main/MainActivity.java
package org.dhis2.usescases.main;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.transition.ChangeBounds;
import android.transition.Transition;
import android.transition.TransitionManager;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ObservableInt;

import com.andrognito.pinlockview.PinLockListener;
import com.android.dbexporterlibrary.ExporterListener;

import org.dhis2.App;
import org.dhis2.R;
import org.dhis2.databinding.ActivityMainBinding;
import org.dhis2.usescases.about.AboutFragment;
import org.dhis2.usescases.development.DevelopmentActivity;
import org.dhis2.usescases.general.ActivityGlobalAbstract;
import org.dhis2.usescases.general.FragmentGlobalAbstract;
import org.dhis2.usescases.jira.JiraFragment;
import org.dhis2.usescases.main.program.ProgramFragment;
import org.dhis2.usescases.qrReader.QrReaderFragment;
import org.dhis2.usescases.settings.ErrorDialog;
import org.dhis2.usescases.settings.SyncManagerFragment;
import org.dhis2.usescases.teiDashboard.nfc_data.NfcDataWriteActivity;
import org.dhis2.utils.Constants;
import org.dhis2.utils.DateUtils;
import org.dhis2.utils.filters.FilterManager;
import org.dhis2.utils.filters.FiltersAdapter;
import org.hisp.dhis.android.core.imports.TrackerImportConflict;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;

import javax.inject.Inject;

import io.reactivex.functions.Consumer;

import static org.dhis2.utils.analytics.AnalyticsConstants.BLOCK_SESSION;
import static org.dhis2.utils.analytics.AnalyticsConstants.CLICK;
import static org.dhis2.utils.analytics.AnalyticsConstants.CLOSE_SESSION;


public class MainActivity extends ActivityGlobalAbstract implements MainContracts.View, ExporterListener {

    private static final int PERMISSION_REQUEST = 1987;
    public ActivityMainBinding binding;
    @Inject
    MainContracts.Presenter presenter;

    private ProgramFragment programFragment;
    @VisibleForTesting
    protected FragmentGlobalAbstract activeFragment;

    ObservableInt currentFragment = new ObservableInt(R.id.menu_home);
    private boolean isPinLayoutVisible = false;

    private int fragId;
    private SharedPreferences prefs;
    private boolean backDropActive = false;
    private FiltersAdapter adapter;

    //-------------------------------------
    //region LIFECYCLE

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Objects.requireNonNull(((App) getApplicationContext()).userComponent()).plus(new MainModule()).inject(this);

        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        binding.setPresenter(presenter);
        binding.navView.setNavigationItemSelectedListener(item -> {
            changeFragment(item.getItemId());
            return false;
        });
        analyticsHelper.setEvent("sale", "algo", "o que");
        binding.pinLayout.pinLockView.attachIndicatorDots(binding.pinLayout.indicatorDots);
        binding.pinLayout.pinLockView.setPinLockListener(new PinLockListener() {
            @Override
            public void onComplete(String pin) {
                presenter.blockSession(pin);
            }

            @Override
            public void onEmpty() {

            }

            @Override
            public void onPinChange(int pinLength, String intermediatePin) {

            }
        });

        if (savedInstanceState != null) {
            int frag = savedInstanceState.getInt("Fragment");
            currentFragment.set(frag);
            binding.setCurrentFragment(currentFragment);
            changeFragment(frag);
        } else {
            binding.setCurrentFragment(currentFragment);
            changeFragment(R.id.menu_home);
        }

        prefs = getAbstracContext().getSharedPreferences(
                Constants.SHARE_PREFS, Context.MODE_PRIVATE);


        adapter = new FiltersAdapter();
        binding.filterLayout.setAdapter(adapter);

        binding.moreOptions.setOnLongClickListener(v -> {
            startActivity(DevelopmentActivity.class, null, false, false, null);
            return false;
        });
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("Fragment", fragId);
    }

    @Override
    protected void onResume() {
        super.onResume();
        presenter.init(this);

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED
                || ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {

            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.CAMERA},
                    PERMISSION_REQUEST);
        }
        binding.setTotalFilters(FilterManager.getInstance().getTotalFilters());
        adapter.notifyDataSetChanged();
    }

    @Override
    protected void onPause() {
        presenter.onDetach();
        super.onPause();
    }

    //endregion

    /*User info methods*/

    @NonNull
    @Override
    public Consumer<String> renderUsername() {
        return username -> {
            binding.setUserName(username);
            ((TextView) binding.navView.getHeaderView(0).findViewById(R.id.user_info)).setText(username);
            binding.executePendingBindings();
        };
    }

    /*End of user info methods*/

    @Override
    public void openDrawer(int gravity) {
        if (!binding.mainDrawerLayout.isDrawerOpen(gravity))
            binding.mainDrawerLayout.openDrawer(gravity);
        else
            binding.mainDrawerLayout.closeDrawer(gravity);
    }

    @Override
    public void showHideFilter() {
        Transition transition = new ChangeBounds();
        transition.setDuration(200);
        TransitionManager.beginDelayedTransition(binding.backdropLayout, transition);
        backDropActive = !backDropActive;
        ConstraintSet initSet = new ConstraintSet();
        initSet.clone(binding.backdropLayout);
        if (backDropActive)
            initSet.connect(R.id.fragment_container, ConstraintSet.TOP, R.id.filterLayout, ConstraintSet.BOTTOM, 50);
        else
            initSet.connect(R.id.fragment_container, ConstraintSet.TOP, R.id.toolbar, ConstraintSet.BOTTOM, 0);
        initSet.applyTo(binding.backdropLayout);
        programFragment.openFilter(backDropActive);
    }

    private void checkFilterEnabled() {
        binding.filter.setBackgroundColor(programFragment.areFiltersApplied() ? getAccentColor() : getPrimaryColor());
        binding.filter.setColorFilter(programFragment.areFiltersApplied() ? getPrimaryColor() : getAccentColor(), PorterDuff.Mode.SRC_IN);
        binding.filter.setBackgroundResource(programFragment.areFiltersApplied() ? R.drawable.white_circle : 0);
    }

    @Override
    public void onLockClick() {
        if (prefs.getString("pin", null) == null) {
            binding.mainDrawerLayout.closeDrawers();
            binding.pinLayout.getRoot().setVisibility(View.VISIBLE);
            isPinLayoutVisible = true;
        } else
            presenter.blockSession(null);
    }

    @Override
    public void onBackPressed() {
        if (fragId != R.id.menu_home) {
            changeFragment(R.id.menu_home);
        } else if (isPinLayoutVisible) {
            isPinLayoutVisible = false;
            startActivity(new Intent(MainActivity.this, MainActivity.class));
            finish();
        } else
            super.onBackPressed();
    }

    @Override
    public void changeFragment(int id) {
        fragId = id;
        binding.navView.setCheckedItem(id);
        activeFragment = null;
        String tag = null;

        switch (id) {
            case R.id.sync_manager:
                activeFragment = new SyncManagerFragment();
                tag = getString(R.string.SYNC_MANAGER);
                binding.filter.setVisibility(View.GONE);
                break;
            case R.id.qr_scan:
                activeFragment = new QrReaderFragment();
                tag = getString(R.string.QR_SCANNER);
                binding.filter.setVisibility(View.GONE);
                break;
            case R.id.nfc_scan:
                Intent intentNfc = new Intent(this, NfcDataWriteActivity.class);
                startActivity(intentNfc);
                break;
            case R.id.menu_jira:
                activeFragment = new JiraFragment();
                tag = getString(R.string.jira_report);
                binding.filter.setVisibility(View.GONE);
                break;
            case R.id.menu_about:
                activeFragment = new AboutFragment();
                tag = getString(R.string.about);
                binding.filter.setVisibility(View.GONE);
                break;
            case R.id.block_button:
                analyticsHelper.setEvent(BLOCK_SESSION, CLICK, BLOCK_SESSION);
                onLockClick();
                break;
            case R.id.logout_button:
                analyticsHelper.setEvent(CLOSE_SESSION, CLICK, CLOSE_SESSION);
                presenter.logOut();
                break;
            case R.id.menu_home:
            default:
                activeFragment = new ProgramFragment();
                programFragment = (ProgramFragment) activeFragment;
                tag = getString(R.string.done_task);
                binding.filter.setVisibility(View.VISIBLE);
                break;
        }

        if (activeFragment != null) {
            currentFragment.set(id);
            getSupportFragmentManager().beginTransaction().replace(R.id.fragment_container, activeFragment, tag).commitAllowingStateLoss();
            binding.title.setText(tag);
        }
        binding.mainDrawerLayout.closeDrawers();

        if (backDropActive && !(activeFragment instanceof ProgramFragment))
            showHideFilter();

    }

    @Override
    public void showSyncErrors(List<TrackerImportConflict> data) {
        new ErrorDialog().setData(data).show(getSupportFragmentManager().beginTransaction(), ErrorDialog.TAG);
    }

    @Override
    public void updateFilters(int totalFilters) {
        binding.setTotalFilters(totalFilters);
    }

    @Override
    public void showPeriodRequest(FilterManager.PeriodRequest periodRequest) {
        if (periodRequest == FilterManager.PeriodRequest.FROM_TO) {
            DateUtils.getInstance().showFromToSelector(this, FilterManager.getInstance()::addPeriod);
        } else {
            DateUtils.getInstance().showPeriodDialog(this, datePeriods -> {
                        FilterManager.getInstance().addPeriod(datePeriods);
                    },
                    true);
        }
    }

    public void setTitle(String title) {
        binding.title.setText(title);
    }

    public FiltersAdapter getAdapter() {
        return adapter;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void showTutorial(boolean shaked) {
        switch (fragId) {
            case R.id.menu_home:
                ((ProgramFragment) activeFragment).setTutorial();
                break;
            case R.id.sync_manager:
                ((SyncManagerFragment) activeFragment).showTutorial();
                break;
            default:
                showToast(getString(R.string.no_intructions));
                break;
        }
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == FilterManager.OU_TREE && resultCode == Activity.RESULT_OK) {
            adapter.notifyDataSetChanged();
            updateFilters(FilterManager.getInstance().getTotalFilters());
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void fail(@NotNull String message, @NotNull String exception) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void success(@NotNull String s) {
        Toast.makeText(this, s, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/org/dhis2/usescases/main/MainActivity.kt
package org.dhis2.usescases.main

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.transition.ChangeBounds
import android.transition.TransitionManager
import android.view.View
import android.webkit.MimeTypeMap
import android.widget.TextView
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.constraintlayout.widget.ConstraintSet
import androidx.core.app.NotificationCompat
import androidx.core.net.toUri
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.lifecycleScope
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.launch
import org.dhis2.BuildConfig
import org.dhis2.R
import org.dhis2.bindings.hasPermissions
import org.dhis2.commons.filters.FilterManager
import org.dhis2.commons.filters.Filters
import org.dhis2.commons.filters.FiltersAdapter
import org.dhis2.commons.filters.periods.ui.FilterPeriodsDialog
import org.dhis2.commons.filters.periods.ui.FilterPeriodsDialog.Companion.FILTER_DIALOG
import org.dhis2.commons.orgunitselector.OUTreeFragment
import org.dhis2.commons.sync.OnDismissListener
import org.dhis2.commons.sync.SyncContext
import org.dhis2.databinding.ActivityMainBinding
import org.dhis2.usescases.development.DevelopmentActivity
import org.dhis2.usescases.general.ActivityGlobalAbstract
import org.dhis2.usescases.login.LoginActivity
import org.dhis2.usescases.main.ui.NewVersionDialog
import org.dhis2.usescases.main.ui.TAG
import org.dhis2.usescases.main.ui.model.HomeEffect
import org.dhis2.usescases.main.ui.model.HomeScreenState
import org.dhis2.usescases.main.ui.model.VersionToUpdateState
import org.dhis2.usescases.main.ui.screens.MainScreen
import org.dhis2.utils.granularsync.SyncStatusDialog
import org.hisp.dhis.mobile.ui.designsystem.theme.DHIS2Theme
import org.dhis2.mobile.plugin.domain.LoadPluginsUseCase
import org.koin.android.ext.android.inject
import org.koin.androidx.viewmodel.ext.android.viewModel
import org.koin.core.parameter.parametersOf
import kotlin.getValue

private const val FRAGMENT = "Fragment"
private const val INIT_DATA_SYNC = "INIT_DATA_SYNC"
private const val WIPE_NOTIFICATION = "wipe_notification"
private const val RESTART = "Restart"
const val AVOID_SYNC = "AvoidSync"

class MainActivity : ActivityGlobalAbstract() {
    private lateinit var binding: ActivityMainBinding
    private val mainViewModel: MainViewModel by viewModel(parameters = {
        parametersOf(
            context,
            supportFragmentManager,
            intent.getBooleanExtra(AVOID_SYNC, false),
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                intent.getParcelableExtra(FRAGMENT, MainScreenType::class.java)
                    ?: MainScreenType.Home(
                        HomeScreen.Programs
                    )
            } else {
                @Suppress("DEPRECATION")
                intent.getParcelableExtra(FRAGMENT) ?: MainScreenType.Home(HomeScreen.Programs)
            }
        )
    })

    private val filtersAdapter: FiltersAdapter = FiltersAdapter()
    private val loadPluginsUseCase: LoadPluginsUseCase by inject()

    private var backDropActive = false
    private val requestWritePermissions =
        registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R || granted) {
                onDownloadNewVersion()
            } else {
                Toast
                    .makeText(
                        context,
                        getString(R.string.storage_denied),
                        Toast.LENGTH_LONG,
                    ).show()
            }
        }

    private val navigationLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { }

    companion object {
        fun intent(
            context: Context,
            initScreen: MainScreenType? = null,
            launchDataSync: Boolean = false,
        ): Intent =
            Intent(context, MainActivity::class.java).apply {
                putExtras(
                    bundle(
                        initScreen = initScreen,
                        launchDataSync = launchDataSync
                    )
                )
            }

        fun bundle(
            initScreen: MainScreenType? = null,
            launchDataSync: Boolean = false,
        ) = Bundle().apply {
            initScreen?.let {
                putParcelable(FRAGMENT, initScreen)
            }
            putBoolean(INIT_DATA_SYNC, launchDataSync)
        }
    }

    //region LIFECYCLE
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContent {
            DHIS2Theme {
                val screenState by mainViewModel.homeScreenState.collectAsState()
                MainScreen(
                    screenState = screenState,
                    effects = mainViewModel.homeEffects,
                    onAction = mainViewModel::onAction,
                    onEffect = { effect ->
                        when (effect) {
                            HomeEffect.BlockSession -> finish()
                            is HomeEffect.GoToLogin -> goToLogin(
                                accountsCount = effect.accountsCount,
                                isDeletion = effect.isDeletion,
                            )
                            HomeEffect.OrgUnitFilterRequest -> openOrgUnitTreeSelector()
                            is HomeEffect.PeriodFilterRequest -> showPeriodRequest(effect.periodRequest)
                            HomeEffect.ShowDeleteNotification -> showProgressDeleteNotification()
                            HomeEffect.ShowGranularSync -> showGranularSync()
                            is HomeEffect.SingleProgramNavigation -> navigationLauncher.navigateTo(
                                this@MainActivity,
                                effect.homeItemData,
                            )
                            HomeEffect.ShowPinDialog -> { /*handled in composable*/ }
                            HomeEffect.ToggleSideMenu -> openDrawer()
                            HomeEffect.ToggleFilters -> showHideFilter()
                        }
                    },
                    onNewState = { state ->
                        if (::binding.isInitialized) {
                            updateScreen(state)
                        }
                    },
                    onNewScreen = { currentScreen ->
                        if (backDropActive) {
                            showHideFilter()
                        }
                        val navigationId = when (currentScreen) {
                            MainScreenType.About -> R.id.menu_about
                            is MainScreenType.Home -> R.id.menu_home
                            MainScreenType.Loading -> null
                            MainScreenType.QRScanner -> R.id.qr_scan
                            MainScreenType.Settings -> R.id.sync_manager
                            MainScreenType.TroubleShooting -> R.id.menu_troubleshooting
                        }
                        navigationId?.let {
                            changeFragment(it)
                            initCurrentScreen()
                        }
                    },
                    onLayoutInflated = {
                        binding = it
                        initBinding()
                    },
                )
            }
        }
    }

    private fun initBinding() {
        binding.navView.setNavigationItemSelectedListener { item ->
            changeFragment(item.itemId)
            false
        }

        binding.mainDrawerLayout.addDrawerListener(object : DrawerLayout.DrawerListener {
            override fun onDrawerSlide(drawerView: View, slideOffset: Float) = Unit

            override fun onDrawerOpened(drawerView: View) = Unit

            override fun onDrawerClosed(drawerView: View) {
                initCurrentScreen()
            }

            override fun onDrawerStateChanged(newState: Int) = Unit
        })

        binding.filterRecycler.adapter = filtersAdapter

        if (BuildConfig.DEBUG || BuildConfig.FLAVOR == "dhis2Training") {
            binding.navView.menu
                .findItem(R.id.menu_troubleshooting)
                .isVisible = true
            binding.navView.menu
                .findItem(R.id.menu_dev)
                .isVisible = true
        }

        lifecycleScope.launch { loadPluginsUseCase(Unit) }

        checkNotificationPermission()
    }

    private fun updateScreen(screenState: HomeScreenState) {
        binding.navView.getHeaderView(0).findViewById<TextView>(R.id.user_info)
            .text = screenState.userName

        if (screenState.homeFilters.isNotEmpty()) {
            filtersAdapter.submitList(screenState.homeFilters)
        }

        when (val versionState = screenState.versionToUpdate) {
            VersionToUpdateState.Downloading ->
                binding.toolbarProgress.show()

            is VersionToUpdateState.New ->
                if (supportFragmentManager.findFragmentByTag(TAG) == null) {
                    showNewVersionAlert(versionState.version)
                }

            VersionToUpdateState.None ->
                binding.toolbarProgress.hide()
        }
    }

    /*TODO: MOVE TO BE HANDLED AFTER LOGIN*/
    private fun checkNotificationPermission() {
        if (!hasPermissions(arrayOf(Manifest.permission.POST_NOTIFICATIONS)) and
            (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
        ) {
            requestNotificationPermission.launch(Manifest.permission.POST_NOTIFICATIONS)
        }
    }

    private fun showHideFilter() {
        val transition = ChangeBounds()
        transition.duration = 200
        TransitionManager.beginDelayedTransition(binding.backdropLayout, transition)
        backDropActive = !backDropActive
        val initSet = ConstraintSet()
        initSet.clone(binding.backdropLayout)
        if (backDropActive) {
            initSet.connect(
                R.id.fragment_container,
                ConstraintSet.TOP,
                R.id.filterRecycler,
                ConstraintSet.BOTTOM,
                50,
            )
            mainViewModel.updateNavigationBarVisibility(false)
        } else {
            initSet.connect(
                R.id.fragment_container,
                ConstraintSet.TOP,
                R.id.toolbarProgress,
                ConstraintSet.BOTTOM,
                0,
            )
            mainViewModel.updateNavigationBarVisibility(true)
        }
        initSet.applyTo(binding.backdropLayout)
    }

    private fun showGranularSync() {
        SyncStatusDialog
            .Builder()
            .withContext(this)
            .withSyncContext(SyncContext.Global())
            .onDismissListener(
                object : OnDismissListener {
                    override fun onDismiss(hasChanged: Boolean) {
                        mainViewModel.onGranularSyncFinished(hasChanged)
                    }
                },
            ).onNoConnectionListener {
                val contextView = findViewById<View>(R.id.navigationBar)
                Snackbar
                    .make(
                        contextView,
                        R.string.sync_offline_check_connection,
                        Snackbar.LENGTH_SHORT,
                    ).show()
            }.show("ALL_SYNC")
    }

    private fun showPeriodRequest(periodRequest: FilterManager.PeriodRequest) {
        if (periodRequest == FilterManager.PeriodRequest.FROM_TO) {
            FilterPeriodsDialog
                .newPeriodsFilter(filterType = Filters.PERIOD, isFromToFilter = true)
                .show(supportFragmentManager, FILTER_DIALOG)
        } else {
            FilterPeriodsDialog
                .newPeriodsFilter(filterType = Filters.PERIOD)
                .show(supportFragmentManager, FILTER_DIALOG)
        }
    }

    private fun openOrgUnitTreeSelector() {
        OUTreeFragment
            .Builder()
            .withPreselectedOrgUnits(
                FilterManager
                    .getInstance()
                    .orgUnitFilters
                    .map { it.uid() }
                    .toMutableList(),
            ).onSelection(mainViewModel::setOrgUnitFilters)
            .build()
            .show(supportFragmentManager, "OUTreeFragment")
    }

    private fun goToLogin(
        accountsCount: Int,
        isDeletion: Boolean,
    ) {
        startActivity(
            LoginActivity::class.java,
            LoginActivity.bundle(
                accountsCount = accountsCount,
                isDeletion = isDeletion,
                fromMainActivity = true,
            ),
            finishCurrent = true,
            finishAll = true,
            transition = null,
        )
    }

    private fun openDrawer() {
        if (!binding.mainDrawerLayout.isDrawerOpen(GravityCompat.START)) {
            binding.mainDrawerLayout.openDrawer(GravityCompat.START)
        } else {
            binding.mainDrawerLayout.closeDrawer(GravityCompat.START)
        }
    }

    private fun changeFragment(id: Int) {
        binding.navView.setCheckedItem(id)
        binding.mainDrawerLayout.closeDrawers()
    }

    private fun initCurrentScreen() {
        when (binding.navView.checkedItem?.itemId) {
            R.id.sync_manager ->
                mainViewModel.onChangeScreen(MainScreenType.Settings)

            R.id.qr_scan ->
                mainViewModel.onChangeScreen(MainScreenType.QRScanner)

            R.id.menu_about ->
                mainViewModel.onChangeScreen(MainScreenType.About)

            R.id.block_button ->
                mainViewModel.onBlockSession()

            R.id.logout_button ->
                mainViewModel.logOut()

            R.id.menu_home ->
                mainViewModel.onChangeToHome()

            R.id.menu_troubleshooting ->
                mainViewModel.onChangeScreen(MainScreenType.TroubleShooting)

            R.id.menu_dev ->
                startActivity(Intent(this, DevelopmentActivity::class.java))

            R.id.delete_account ->
                confirmAccountDelete()

        }
    }

    private fun confirmAccountDelete() {
        MaterialAlertDialogBuilder(this, R.style.MaterialDialog)
            .setTitle(getString(R.string.delete_account))
            .setMessage(getString(R.string.wipe_data_meesage))
            .setView(R.layout.warning_layout)
            .setPositiveButton(getString(R.string.wipe_data_ok)) { _, _ -> mainViewModel.onDeleteAccount() }
            .setNegativeButton(getString(R.string.wipe_data_no)) { dialog, _ ->
                dialog.dismiss()
            }.show()
    }

    private fun showProgressDeleteNotification() {
        val notificationManager =
            context.getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val mChannel =
                NotificationChannel(
                    WIPE_NOTIFICATION,
                    RESTART,
                    NotificationManager.IMPORTANCE_HIGH,
                )
            notificationManager.createNotificationChannel(mChannel)
        }
        val notificationBuilder =
            NotificationCompat
                .Builder(context, WIPE_NOTIFICATION)
                .setSmallIcon(R.drawable.ic_sync)
                .setContentTitle(getString(R.string.wipe_data))
                .setContentText(getString(R.string.please_wait))
                .setOngoing(true)
                .setAutoCancel(false)
                .setPriority(NotificationCompat.PRIORITY_HIGH)

        notificationManager.notify(123456, notificationBuilder.build())
    }

    private fun showNewVersionAlert(version: String) {
        NewVersionDialog(
            newVersion = version,
            onRemindMeLater = mainViewModel::remindLaterAlertNewVersion,
            onDownloadVersion = {
                if (hasPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE))) {
                    onDownloadNewVersion()
                } else {
                    requestWritePermissions.launch(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                }
            },
        ).show(supportFragmentManager)
    }

    private fun onDownloadNewVersion() {
        mainViewModel.downloadVersion(
            onDownloadCompleted = ::installAPK,
            onLaunchUrl = ::launchUrl,
        )
    }

    private fun installAPK(apkUri: Uri) {
        when {
            hasNoPermissionToInstall() ->
                manageUnknownSources.launch(
                    Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES)
                        .setData(String.format("package:%s", packageName).toUri()),
                )

            !hasPermissions(arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)) && Build.VERSION.SDK_INT < Build.VERSION_CODES.R ->
                requestReadStoragePermission.launch(Manifest.permission.READ_EXTERNAL_STORAGE)

            else ->
                Intent(Intent.ACTION_VIEW).apply {
                    val mime = MimeTypeMap.getSingleton()
                    val ext = apkUri.path?.substringAfterLast(("."))
                    val type: String? = mime.getMimeTypeFromExtension(ext)
                    addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                    setDataAndType(apkUri, type)
                    startActivity(this)
                }
        }
    }

    private fun hasNoPermissionToInstall(): Boolean =
        Build.VERSION.SDK_INT >= Build.VERSION_CODES.O &&
                !packageManager.canRequestPackageInstalls()

    private val manageUnknownSources =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            if (hasNoPermissionToInstall()) {
                Toast
                    .makeText(
                        context,
                        getString(R.string.unknow_sources_denied),
                        Toast.LENGTH_LONG,
                    ).show()
            } else {
                onDownloadNewVersion()
            }
        }

    private val requestReadStoragePermission =
        registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                onDownloadNewVersion()
            } else if (granted) {
                onDownloadNewVersion()
            } else {
                Toast
                    .makeText(
                        context,
                        getString(R.string.storage_denied),
                        Toast.LENGTH_LONG,
                    ).show()
            }
        }

    private val requestNotificationPermission =
        registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
            val message = if (granted) {
                getString(R.string.permission_notification_granted)
            } else {
                getString(R.string.permission_notification_denied)
            }

            Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
        }

    private fun launchUrl(uri: Uri) {
        val intent = Intent(Intent.ACTION_VIEW, uri)
        startActivity(intent)
    }

    /** Required by QrReaderFragment. It should retrieve the activity mainViewModel and call
     * onChangeToHome directly
     * */
    fun goToHome() {
        mainViewModel.onChangeToHome()
    }
}
